Log subproceses stdout and stderr in temp files

MVrachev opened the pull request:
**Fixes issues #**: https://github.com/theupdateframework/tuf/issues/1039, https://github.com/theupdateframework/tuf/issues/1119

**Description of the changes being introduced by the pull request**:

**This pr replaces pr https://github.com/theupdateframework/tuf/pull/1041.**

Logging the stdout and stderr from the test subprocesses into
temporary files clean the console from unnecessary messages from
the server-side such as "code 404, message File not found" or
"GET" queries.

The unit tests are executed in sequential order and that's why
we can reuse one temporary file object for multiple tests.

I have used this pr to remove some redundant functions, comments, and variables.

For this pr I received great help from @jku!
Thanks!

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

MVrachev commented:
On a successful test run the console logs look like this: 
![image](https://user-images.githubusercontent.com/16246778/89891691-31073500-dbde-11ea-8319-87764134ad15.png)

I wanted to simulate an error in the subprocess and I changed the name of the server script from `simple_server.py` to `tsimple_server.py` and here how the error was logged: 
![image](https://user-images.githubusercontent.com/16246778/89891862-79265780-dbde-11ea-8203-71a8112e5c0d.png)

jku commented:
Two comments from a quick test:
* the child process output does get buffered for me (so `python3 test_updater.py TestUpdater.test_4_refresh` does not notice the 404s): we should start the server using  "python -u simple_server.py" to prevent buffering
* the server output should not be warning because warnings are printed by default. Also we don't really know what the messages are and mostly do not care. info or debug level maybe?

MVrachev commented:
> * the child process output does get buffered for me (so `python3 test_updater.py TestUpdater.test_4_refresh` does not notice the 404s): we should start the server using  "python -u simple_server.py" to prevent buffering

Yes, I checked that and you are right.
The child process buffers.

The solution that I propose is to use **_two temporary files_**: one for stdout and one for stderr.
I imagine that there could be an option which will give you control if you want to see the stdout or not.
I propose that by the default, if the stderr temp file has any information in it, then we would print it and the stdout file will be ignored.


> * the server output should not be a warning because warnings are printed by default. Also, we don't really know what the messages are and mostly do not care. info or debug level maybe?

If we separate stderr from stdout as I proposed above, then we can use `logger.warning()` or `logger.error()` for stderr and `logger.debug()` or `logger.info()` for stdout.
My opinion is that `logger.error()` is appropriate for `stderr` and `logger.debug()` for stdout.

My bigger concern with that approach is how we are going to propagate the cmd option when using tox or even calling the tests with `python aggregate_tests.py`...

jku commented:
> The solution that I propose is to use two temporary files: one for stdout and one for stderr.

You can do that but I don't think it's necessarily worth the effort: 
* what simple_server.py considers stderr/stdout might not map to our categories anyway (as an example: we might expect the server to print errors in some situations: then it's not an error for us)
* all of simple_server.py output is mostly debugging aid anyway: We'll want to look at it only when something isn't going the way we expected
* if you separate the streams, you no longer know if the stderr line came before the stdout line...

I think dumping all the server output to debug or info is fine

jku commented:
You know what: we should think about combining (or re-ordering) this PR and issue #1111 .

If we had this in utils.py (edited after chat with Martin):

```
class TestServerProcess():
 
  # starts server, sets up the TempFile, waits for server response before returning
  def __init__(self, server='simple_server.py', port, extra_args=None, timeout=10)
  
  # logs contents of TempFile with given logger, truncates buffer
  def flush_log(self, logger)
  
  # handles kill() and communicate() on the subprocess
  def kill(self)
```
 I think this PR could be vastly smaller and easier to modify later, AND we would have the infra needed by #1111

Commit:
Remove redundant "url" class member from tests

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I am working to create an abstraction that will handle the subprocess and the temporary files used for logging.
While I am doing this I will release a couple of commits and I will mark this pr as a draft before it's ready for full review.

Commit:
Log subproceses stdout and stderr in temp files

Logging the stdout and stderr from the test subprocesses into
temporary files clean the console from unnecessary messages from
the server-side such as "code 404, message File not found" or
"GET" queries.

I have decided to create TestServerProcess class that will handle
the server subprocess creation and redirection to a temporary file
object. That way that code can be reused in more than 10 files.

Also, I have cleaned some parts of the unit test to make them more
readable and efficient with the new abstraction.

The unit tests are executed in sequential order and that's why
we can reuse one temporary file object for multiple tests.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I have pushed the initial set of commits in which I created the class which handles the server subprocess and the temporary files used for logging.

I have decided to use `logging.info` for logging in the `flus_log` function because stdout and stderr are logged together and thus it seems like a reasonable default value.

On a successful test right now no logs like `code 404, message File not found` or "GET" queries are reported.

I am planning to push at least 1 commit more in which I will remove or reword some of the test comments.

jku commented on tests/test_download.py:
```diff
@@ -260,27 +259,30 @@ def test_https_connection(self):
+    good_https_url = 'https://localhost:' + str(port1) + suffix
+    good2_https_url = 'https://localhost:' + str(port2) + suffix
+    bad_https_url = 'https://localhost:' + str(port3) + suffix
+    expired_https_url = 'https://localhost:' + str(port3) + suffix
```
port number typo -- if this did not come up in tests, maybe check why

MVrachev replied:
It is not shown in the tests because we are checking that there is an SSL error and not what exact type is:
```
with self.assertRaises(requests.exceptions.SSLError):
        download.unsafe_download(expired_https_url, target_data_length)
```

jku commented on tests/utils.py:
```diff
@@ -74,3 +77,118 @@ def wait_for_server(host, port, timeout=10):
+    self.__temp_log_file.seek(0)
+    log_message = self.__temp_log_file.read()
+
+    if len(log_message) > 0:
+      self.__logger.info("*******************")
+      # If you run only one test file with "python <file_name>"
+      # you will receive unnecessary "(null)" and "__main__."
+      test_name = test_name.replace("__main__.", "")
+      log_message = log_message.replace("(null): ", "")
+
+      self.__logger.info("Logs from the server subprocess!")
+      if test_name:
+        self.__logger.info("Logs for test " + test_name)
+
+      self.__logger.info("INFO: " + log_message)
+      self.__logger.info("*******************")
+      # Make sure the file is empty before the next test logs new information.
+      self.__temp_log_file.truncate(0)
+
```
We already discussed but I'll leave my opinion here as well for transparency: I think this does things it does not need to... 
* We don't print test names in any other logging either: if it's not clear which test is running, why not improve that in general instead of just for server output
* if the log message from the server does not look good, I think we should fix the test server output instead of replacing things in the output here
* the output missing the server name however: e.g. "simple_server.py": this would be important sometimes

these are opinions though, feel free to disregard

MVrachev replied:
> We don't print test names in any other logging either: if it's not clear which test is running, why not improve that in general instead of just for server output

I think the `test_name` argument can be removed because as we discussed on a test failure the logs are showed once before all tests and a second time just after the test name:
![image](https://user-images.githubusercontent.com/16246778/92588350-86983580-f2a1-11ea-9537-76de15931b4f.png)

I don't think we can add the test names to all logs because many of the logs are coming from public API functions and there is no need to add `test_name` argument just for that.

> if the log message from the server does not look good, I think we should fix the test server output instead of replacing things in the output here

Well, yes, maybe you are right. I will remove that part.

> the output missing the server name however: e.g. "simple_server.py": this would be important sometimes

Will add that.

jku commented on tests/test_download.py:
```diff
@@ -260,10 +260,13 @@ def test_https_connection(self):
+    port1 = random.randint(30000, 35000)
+    port2 = random.randint(35001, 40000)
+    port3 = random.randint(40001, 45000)
+    port4 = random.randint(45001, 50000)
```
Hmmm, this now makes comment on line 261 untrue I think

MVrachev replied:
I am planning to add a function `_generate_port()` in the TestServerProcess class where I will add a function that generates a port and checks if the port is busy.
So, this code will be removed.

jku commented:
Looks good, thanks! I'll take this for a test run later to see what the output looks like

MVrachev commented:
I don't know why but when I run `python test_arbitrary_package_attack.py` I get a resource warning even though I have closed the `temporary_log_file` in the `clean` function: 
`/home/mvrachev/src/tuf/.editable_both/lib/python3.6/posixpath.py:86: ResourceWarning: unclosed file <_io.FileIO name=5 mode='rb+' closefd=True>
  for b in map(os.fspath, p):`

Same for every file where I am using this new abstraction...

jku commented:
> I don't know why but when I run `python test_arbitrary_package_attack.py` I get a resource warning even though I have closed the `temporary_log_file` in the `clean` function:
> `/home/mvrachev/src/tuf/.editable_both/lib/python3.6/posixpath.py:86: ResourceWarning: unclosed file <_io.FileIO name=5 mode='rb+' closefd=True> for b in map(os.fspath, p):`

I'm guessing this is not _your_ unclosed file -- updater currently leaves unclosed files on pretty much every "error" situation I think. Maybe check if same warning appears on develop branch? If it does then you don't need to handle it here.

You can also get more details on the warning with` PYTHONTRACEMALLOC=50` if you want

Commit:
Remove unneceserry checks in server files

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
> > I don't know why but when I run `python test_arbitrary_package_attack.py` I get a resource warning even though I have closed the `temporary_log_file` in the `clean` function:
> > `/home/mvrachev/src/tuf/.editable_both/lib/python3.6/posixpath.py:86: ResourceWarning: unclosed file <_io.FileIO name=5 mode='rb+' closefd=True> for b in map(os.fspath, p):`
> 
> I'm guessing this is not _your_ unclosed file -- updater currently leaves unclosed files on pretty much every "error" situation I think. Maybe check if same warning appears on develop branch? If it does then you don't need to handle it here.
> 
> You can also get more details on the warning with` PYTHONTRACEMALLOC=50` if you want


Yes, those warnings are visible in the develop branch too.

Commit:
Remove a not used function

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Remove redundant comments from test classes

I don't see a need to leave a comment about what setupClass,
tearDownClass, setup and tearDown functions do.
There is documentation that describes that.
Additionally, the links referenced in the comments are from
Python 2 is deprecated.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tests/utils.py:
```diff
@@ -74,3 +77,117 @@ def wait_for_server(host, port, timeout=10):
+    log_message = self.__temp_log_file.read()
+
+    if len(log_message) > 0:
+      self.__logger.info("*******************")
+
+      self.__logger.info("Logs from the subprocess running server: " \
+          + self.server)
+
+      messages = log_message.split('\n')
+      for line in messages:
+        if line != "":
+          self.__logger.info("INFO: " + line)
+
+      self.__logger.info("*******************")
```
I still have issues with this bit:
* very verbose: most of the time the log itself is a single line: this prints 4 times as much
* should not print "INFO: ": this is what the log level is for. If the user wants it printed, they will use a log formatter that prints the log level
* I think the earlier version of a single log entry for the server log was better (it can still be multiline) as that can be made to clearly show the beginning and end of server log without need of extra lines of "************"  (I'll paste the example I tested with below) -- do you see a problem with this approach?

```
    # Build a single log entry with multiline content
    title = "Test server (" + self.server + ") output:"
    message = [title] + log_message.splitlines()
    self.__logger.info('\n| '.join(message))
```

This way the normal oneliner now looks like this (assuming basicConfig default formatter)

```
INFO:logger_name:Test server (simple_server.py) output:
| code 404, message File not found
```

MVrachev replied:
I fixed it in an upcoming commit and removed the commit where I introduce the `INFO` prefix before each line.

jku commented on tests/test_proxy_use.py:
```diff
@@ -78,25 +75,21 @@ def setUpClass(cls):
-    cls.https_server_proc = popen_python(
-        ['simple_https_server.py', str(cls.https_port)])
-
+    cls.https_server_handler = utils.TestServerProcess(log=logger,
+        server='simple_https_server.py')
```
This has the same issue as `test_https_connection()` -- there is now a (very small) chance of collision with cls.http_server_handler.port.

I think the real solution is to make simple_server actually report back that it did succeed in binding the port (so TestServerProcess can retry if it did not). It should probably (?) not be part of this PR though: you'll have to choose an interim solution here: I think I'd use `port=cls.http_server_handler.port + 1` because that is closest to current functionality

jku replied:
this applies to all servers below as well

MVrachev replied:
Yes, I won't fix this in this pr, but I will return it to the previous state in both `test_proxy_use.py` and `test_download.py`.

jku commented on tests/utils.py:
```diff
@@ -74,3 +77,117 @@ def wait_for_server(host, port, timeout=10):
+    self.__logger.info('Serving on port: ' + str(self.port))
+
+    if timeout > 0:
+      try:
+        wait_for_server('localhost', self.port, timeout)
+      except Exception as e:
+        self.__logger.error(
+            "Error while waiting for the server to start: " + repr(e))
```
Not sure logging an extra error is useful? if the TimeoutError lacks detail (which I think it does, sorry about that) we should add an informative message to the error where its raised maybe.

MVrachev replied:
I created a new commit where I remove that line and make `TimeoutError` message more comprehensive.

jku commented:
I left a few comments, otherwise I think it looks good

trishankatdatadog commented:
Just wanted to say: thanks for doing all this tedious but necessary work! 🥇

Commit:
Make TimeoutError message more comprehensive

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
Together with the great help of @jku, we have been elaborating over those changes in the last couple of weeks and now I think this pr is ready for a review!

I have rebased and cleared the commit history to make more sense for a newcomer.

This pr hides the logs from the server because it's using `logger.info()` to log information from the stdout and stderr of the server subprocesses. That's why it's better to merge this pr after https://github.com/theupdateframework/tuf/pull/1145 which will provide a way to receive logs with `info` logging level. This shouldn't be a problem to receive a review.

joshuagl commented on tests/test_download.py:
```diff
@@ -73,15 +71,11 @@ def setUp(self):
 
-    utils.wait_for_server('localhost', self.PORT)
+    rel_target_filepath = os.path.basename(target_filepath)
+    self.url = 'http://localhost:' \
+        + str(self.server_process_handler.port) + '/' + rel_target_filepath
```
Nice. 

I'd guess we don't even need to store the url during setup in many tests now that we can get the port number from the `server_process_handler`? 👍

MVrachev replied:
Well, 3 out of 4 tests in `test_download.py` are using the `self.url.` 
That's why it make sense to keep it as a variable.

joshuagl commented on tests/test_proxy_use.py:
```diff
@@ -77,26 +75,24 @@ def setUpClass(cls):
                                 " (proxy_server.py is Python2 only)")
 
-    # Launch a simple HTTP server (serves files in the current dir).
     cls.http_port = random.randint(30000, 45000)
```
Why not just let `TestServerProcess` generate this port number?

MVrachev replied:
It could, I just tried to make the lines shorter and just doesn't change it.
We are planning to leave all instances of port generation outside `tests/utils.py` and it's in the best interest of us if we do that as well.
I will change it.

joshuagl commented on tests/test_updater.py:
```diff
@@ -2065,7 +2044,6 @@ def test_get_valid_targetinfo(self):
     shutil.rmtree(os.path.join(self.repository_directory2, 'metadata'))
     shutil.copytree(os.path.join(self.repository_directory2, 'metadata.staged'),
         os.path.join(self.repository_directory2, 'metadata'))
-
```
Nit: I'm curious why this empty line was removed?

MVrachev replied:
Honestly, it seems it's a mistake I did while I was resolving prs.
I will fix it right away.

joshuagl commented on tests/utils.py:
```diff
@@ -74,3 +77,111 @@ def wait_for_server(host, port, timeout=10):
+
+
+
+  # Logs contents from TempFile, truncates buffer
+  def flush_log(self):
```
Nit: here, and throughout `TestServerProcess`, I think this kind of comment might be more idiomatic as a docstring.
```suggestion
  def flush_log(self):
    """Logs contents from TempFile, truncates buffer"""
```

joshuagl reviewed (approved):
Thanks for this great set of cleanups @MVrachev. I added a few minor comments and questions, but this is pretty much ready to merge. Excellent PR.

MVrachev commented:
I rebased my changes upon the latest changes in develop and changed the second commit `Log subproceses stdout and stderr in temp files` with the comments from Joshua.

Commit:
Reorder the tearDownClass cleanup

Fixes an issue where rmtree tries to access and consequently remove
a temp folder where the server has opened a file already.
This results in error:
"PermissionError: [WinError 32] The process cannot access the file
because it is being used by another process"

For reference read:
https://github.com/theupdateframework/tuf/issues/1119

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I have decided that because I am working in the same code area I will use this pr to fix the issue https://github.com/theupdateframework/tuf/issues/1119 too and that's what the last commit is all about.

jku reviewed (approved):
LGTM. 

Maybe you are aware but I'll mention for future commits: instead of 

> For reference read:
> #1119

you can do `Fixes #1119` (or use a few other keywords) to auto-close on merge -- obviously it's useful only if you know the issue should be closed when commit is merged.

Also, the format `Fixes issue #N` does not seem to work -- this is why this PR cover letter is probably not going to trigger auto-closing either.

MVrachev commented:
> LGTM.
> 
> Maybe you are aware but I'll mention for future commits: instead of
> 
> > For reference read:
> > #1119
> 
> you can do `Fixes #1119` (or use a few other keywords) to auto-close on merge -- obviously it's useful only if you know the issue should be closed when commit is merged.
> 
> Also, the format `Fixes issue #N` does not seem to work -- this is why this PR cover letter is probably not going to trigger auto-closing either.

Thanks! 
That's what motivated me to create pr https://github.com/theupdateframework/tuf/pull/1152.

joshuagl commented on tests/utils.py:
```diff
@@ -97,3 +100,112 @@ def configure_test_logging(argv):
+        Logger which will be used for logging.
+
+      server:
+        Path to the server which will be runned by the subprocess.
```
```suggestion
        Path to the server to run in the subprocess.
```
Runned is not a word, perhaps rephrase like this?

joshuagl reviewed (approved):
This is an excellent set of changes, thank you. I'd like to see a minor fixup of a docstring. After that, I think this is good to merge.

joshuagl commented:
Looks good, thanks Martin. Could you rebase and squash/fixup b9ef099 into e7a8a78?

MVrachev commented:
Updated.

joshuagl commented:
Awesome work, thanks @MVrachev
